library/alloc/src/rc.rs

```diff
@@ -1841,33 +1855,20 @@ impl<T> Weak<T> {
     /// [`new`]: Weak::new
     #[stable(feature = "weak_into_raw", since = "1.45.0")]
     pub unsafe fn from_raw(ptr: *const T) -> Self {
-        if ptr.is_null() {
-            Self::new()
-        } else {
-            // See Rc::from_raw for details
-            unsafe {
-                let offset = data_offset(ptr);
-                let fake_ptr = ptr as *mut RcBox<T>;
-                let ptr = set_data_ptr(fake_ptr, (ptr as *mut u8).offset(-offset));
-                Weak { ptr: NonNull::new(ptr).expect("Invalid pointer passed to from_raw") }
-            }
-        }
-    }
-}
+        // SAFETY: data_offset is safe to call, because this pointer originates from a Weak.
+        // See Weak::as_ptr for context on how the input pointer is derived.
+        let offset = unsafe { data_offset(ptr) };
 
-pub(crate) fn is_dangling<T: ?Sized>(ptr: NonNull<T>) -> bool {
-    let address = ptr.as_ptr() as *mut () as usize;
-    address == usize::MAX
-}
+        // Reverse the offset to find the original RcBox.
+        // SAFETY: we use wrapping_offset here because the pointer may be dangling (iff T: Sized).
```

Comment:
"iff" isn't right here... the pointer is not "dangling if and only if `T` is sized". That would mean that `T: Sized` implies a dangling pointer.

I think a plain "if" is correct here.

Reply:
The actual clause is "(may be dangling) iff (T is Sized)", but just regular if is probably fine here either way. I'll change it soon.

-----

In any case, the desired semantics is that (T is Sized) is a necessary but not sufficient precondition for (pointer is dangling). (It is, however, necessary and sufficient for (pointer may dangle).)

I'm too tired at the moment to remember precisely which english is the "correct" way to encode this, probably "only if," if "iff" is not quite right and I felt the reason to use "iff" in the first place.

Just "if" is correct informally though so I'll just change it to that.

Reply:
@CAD97 from what I can see, this change is still outstanding.